src/protocol/modulus_conversion/specialized_mul.rs

```diff
@@ -0,0 +1,424 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::{fabric::Network, Direction, Identity},
+    protocol::{context::ProtocolContext, RecordId},
+    secret_sharing::Replicated,
+};
+
+use serde::{Deserialize, Serialize};
+
+/// A message sent by each helper when they've multiplied their own shares
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct DValue<F> {
+    d: F,
+}
+
+/// A variant of the IKHC multiplication protocol
+/// for use in special cases where many of the shares are known to be zero.
+/// Original IKHC multiplication protocol from:
+/// K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, and B. Pinkas. High-throughput secure AES computation. In WAHC@CCS 2018, pp. 13–24, 2018
+/// Optimizations from: "Adam in Private: Secure and Fast Training of Deep Neural Networks with Adaptive Moment Estimation"
+/// by Nuttapong Attrapadung, Koki Hamada, Dai Ikarashi, Ryo Kikuchi*, Takahiro Matsuda, Ibuki Mishina, Hiraku Morita, and Jacob C. N. Schuldt
+#[derive(Debug)]
+pub struct SpecializedMul {}
+
+impl SpecializedMul {
+    /// A highly specialized variant of the secure multiplication protocol which is only valid
+    /// in the case where 4 of the 6 shares are zero.
+    ///
+    /// This is taken from Appendix F: "Conversion Protocols" from the paper:
+    /// "Adam in Private: Secure and Fast Training of Deep Neural Networks with Adaptive Moment Estimation"
+    /// by Nuttapong Attrapadung, Koki Hamada, Dai Ikarashi, Ryo Kikuchi*, Takahiro Matsuda,
+    /// Ibuki Mishina, Hiraku Morita, and Jacob C. N. Schuldt
+    ///
+    /// This protocol can only be used in the case where:
+    /// Helper 1 has shares (a, 0) and (0, b)
+    /// Helper 2 has shares (0, 0) and (b, 0)
+    /// Helper 3 has shares (0, a) and (0, 0)
+    ///
+    /// In the IKHC multiplication protocol, each helper computes `d_i` and
+    /// sends it to the next helper. But in this case, `d_2` and `d_3` are publicly known to all
+    /// the helper parties and can be replaced with constants, e.g. 0. Therefore, these do not
+    /// need to be sent.
+    ///
+    /// ## Errors
+    /// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+    /// back via the error response
+    #[allow(dead_code)]
+    pub async fn execute_specialized_1<F: Field, N: Network>(
+        ctx: ProtocolContext<'_, N>,
+        record_id: RecordId,
+        a: Replicated<F>,
```

Comment:
we could also keep `a` and `b`  inside the struct and make this method consume self. alternatively this can just be an `async fn` because there is no state

Reply:
I prefer pure functions and no state, so I'm going to try to keep all my protocols just accepting all inputs.

Reply:
Are you suggesting I should just eliminate the struct entirely and simply export a function? That makes sense! I'll do that.